02-intermediate-ruby/exception-handling.md

```diff
@@ -0,0 +1,313 @@
+# Exception Handling In Rails
+
+## Learning Goals
+At the end of this you should be able to:
+
+- Explain how exception handling works
+- Create rescue blocks to "catch" exceptions
+- Define custom exceptions
+- Evaluate when to use exceptions
+
+## The Story So Far
+
+With any program there are things that can go wrong. Maybe a file you are planning to read doesn't exist, your database server is unavailable, or you try to divide something by zero. In these cases, Ruby programs use _exceptions_ to indicate that something has gone wrong.
+
+We've seen exceptions before. If you try and access an undefined variable, Ruby will raise a `NameError`, and we've even raise our own `ArgumentError` exceptions in the Bank Accounts project.
+
+### Producing Exceptions
+
+For a quick reminder of what this looks like, save the following code as `divide_by_zero.rb`
+
+```ruby
+# divide_by_zero.rb
+quotient = 5 / 0
+puts quotient # this code will never run
+```
+
+and run it from the command line
+
+```
+$ ruby divide_by_zero.rb
+divide_by_zero.rb:2:in `/': divided by 0 (ZeroDivisionError)
+	from divide_by_zero.rb:2:in `<main>'
+```
+
+If you need to produce an exception you can use the `raise` keyword
+
+```ruby
+def make_toast(slices)
+  if slices < 1
+    raise ArgumentError.new("Can't make less than 1 slice of toast (asked for #{slices})")
+  end
+end
+```
+
+### Stack Traces
+
+When an exception is raised, it immediately stops the current method, just like `return`. It will then bubble its way up through the program, method by method. If it makes it all the way to the top, Ruby will print out a summary of the exception and a description of what the program was doing when the exception happened, also known as a **stack trace**.
+
+A stack trace contains a ton of useful information, including a list of methods and blocks that Ruby was inside when the exception was raised. Being able to quickly read a complex stack trace is a super useful skill, so let's examine one now.
+
+```ruby
+# breakfast.rb
+def make_toast(slices)
+  if slices < 1
+    raise ArgumentError.new("Can't make less than 1 slice of toast (asked for #{slices})")
+  end
+	# ...probably do some other stuff...
+end
+
+def make_breakfast(type)
+	if type == "eggs and toast"
+		make_toast(-1)
+		make_eggs
+	end
+end
+
+def make_breakfast_banquet
+	10.times do
+		make_breakfast("eggs and toast")
+	end
+end
+
+make_breakfast_banquet
+```
+
+```
+$ ruby breakfast.rb
+breakfast.rb:4:in `make_toast': Can't make less than 1 slice of toast (asked for -1) (ArgumentError)
+	from breakfast.rb:11:in `make_breakfast'
+	from breakfast.rb:18:in `block in make_breakfast_banquet'
+	from breakfast.rb:17:in `times'
+	from breakfast.rb:17:in `make_breakfast_banquet'
+	from breakfast.rb:22:in `<main>'
+	```
```

Comment:
This line being indented messes up all the formatting of the document after this